kart/point_cloud/v1.py

```diff
@@ -84,13 +102,41 @@ def diff_tile(self, other, tile_filter=FeatureKeyFilter.MATCH_ALL, reverse=False
             "tile",
             key_filter=tile_filter,
             key_decoder_method="decode_tile_path",
-            value_decoder_method="get_tile_summary_from_pointer_blob",
+            value_decoder_method="get_tile_summary_promise_from_path",
             reverse=reverse,
         )
 
-    def get_tile_summary_from_pointer_blob(self, tile_pointer_blob):
-        # For now just return the blob contents as a string - it's a reasonable summary.
-        result = pointer_file_to_json(tile_pointer_blob)
-        if "version" in result:
-            del result["version"]
-        return result
+    def get_tile_summary_promise_from_path(self, tile_path):
+        tile_pointer_blob = self.get_blob_at(tile_path)
+        return functools.partial(
+            self.get_tile_summary_from_pointer_blob, tile_pointer_blob
+        )
+
+    def diff_to_wc(self, wc_diff_context, ds_filter=DatasetKeyFilter.MATCH_ALL):
+        ds_diff = DatasetDiff()
+        tile_filter = ds_filter.get("tile", ds_filter.child_type())
+        ds_diff["tile"] = DeltaDiff(self.diff_tile_to_wc(wc_diff_context, tile_filter))
+        return ds_diff
+
+    def diff_tile_to_wc(self, wc_diff_context, tile_filter):
+        deltas = wc_diff_context.workdir_deltas_by_ds_path().get(self.path)
+        if not deltas:
+            return
+
+        def encode_tile_path_relative(tilename):
+            return self.encode_tile_path(tilename, relative=True)
+
+        yield from self.decode_raw_deltas(
```

Comment:
this call could use some explaining overall I think. Maybe a docstring for diff_tile_to_wc or a comment here explaining why the various decoder things are the right things.

I don't love the idea of having 5-6 decoder callbacks as kwargs here, especially because it seems like a PC dataset is likely to use the same 5-6 callbacks every time (is it?) I can't quite follow it

Would a less callback-focused approach would be better, e.g. a DatasetCheckoutStrategy class with some default methods overridden by a PointCloudV1CheckoutStrategy class. Just a suggestion

Reply:
Again, I will tidy this up a bit now.

There is a pattern for generating diffs which lots of datasets will use:
- take some specific subtree of myself and of a similar dataset (ie self.inner_tree / "meta", other.inner_tree "meta")
- run self_subtree.diff_to_tree(other_subtree) to get a pygit2.Diff
- go through all the resulting (insert, update, delete) deltas
- fix up the paths to be proper relative paths again (ie, prepend "meta/" onto them all
- run some transform on the paths to decide what to call each object (eg decode primary key)
- run some transform on the path to get the content of each object (eg, read and decode feature)
- yield as kart deltas

(I think the above looks like good-ish documentation - I might check it in).

So, that's what diff_subtrees does - it takes the key and value transforms that it needs to apply as callbacks. It could use a different callback strategy eg polymorphism but I'm not feeling inspired - I won't change it right now.

That code is quite similar to the sort of thing we need to do to generate a working copy diff for file-based datasets:
- run some magic index-diffing code to get a pygit2.Diff of changes that the user made to the working copy folder
- go through all the resulting (insert, update, delete) deltas
- fix up the paths, maybe. the paths will all be working-copy paths - we might want the old-paths to be fixed up so that they are dataset paths instead. the dataset paths can and do have a different structure to the wc-paths 
- run some transform on all the paths to decide what to call each object (eg just show its filename)
- run some transform on the path to get the content of each object (eg get its hash/size, for now)
- yield as kart deltas

So, I've reused the same code for both by setting lots of callbacks. But I think I can maybe make it clearer if I make a working-copy-diff specific method that gives the callbacks better names, maybe takes less callbacks.
Again, this could be polymorphic and maybe it should be but it's not yet obvious to me how I can make it polymorphic and also definitely better